### PR TITLE
updated code for Job interface to get Job status as per DEP6

### PR DESCRIPTION
**Pull request by AyushVerma2**
updated code for Job interface to get Job status as per DEP6

### CODE REVIEW

File: src/main/java/sg/dex/starfish/Job.java [hunk 1]
```diff
@@ -176,4 +176,37 @@ public default boolean cancel(boolean mayInterruptIfRunning) {
...
+	 * Gets the status of the Job associated with this Operation.
+	 *
+	 * @return The the status of the Job
+	 */
+	Status getStatus();
+
+	/**
+	 * This enum is for different job status
+	 */
+	enum Status {
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/Job.java]: Would prefer a simple String status I think, to be consistent with the other JSON-consuming aspects of Starfish and enable extensibility without breaking code.

File: src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java [hunk 2]
```diff
@@ -30,19 +30,22 @@ public Job invokeAsync(Map<String, Object> params) {
...
 			catch (Throwable t) {
-				future.completeExceptionally(t); // failure
+				future.completeExceptionally(t);
+				memoryJob.setStatus(Job.Status.failed);// failure
 			}
-			assert (future.isDone());
 		});
 
-		MemoryJob memoryJob = MemoryJob.create(future);
+		memoryJob.setStatus(Job.Status.running);
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java]: Race condition? Job could already have completed? Must set the status to running immediately before executing the Job!

File: src/main/java/sg/dex/starfish/impl/memory/MemoryJob.java [hunk 3]
```diff
@@ -70,14 +74,23 @@ public boolean isCancelled() {
 	@Override
 	public Map<String,Object> get(long timeoutMillis, TimeUnit timeUnit) {
 		try {
+			this.setStatus(Status.succeeded);
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/MemoryJob.java]: This looks wrong. Should set status to succeeded in the thread executing the job immediately after success.

File: src/main/java/sg/dex/starfish/impl/memory/MemoryJob.java [hunk 4]
```diff
@@ -70,14 +74,23 @@ public boolean isCancelled() {
...
 			throw new JobFailedException("Job failed with exception: "+cause,e);
 		}
 	}
 
+	@Override
+	public Job.Status getStatus() {
+		return this.status;
+	}
+	public void setStatus( Job.Status status) {
+		this.status = status;
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/MemoryJob.java]: Needs protection. You shouldn't be able to change the status of a "failed" or "succeeded" job, for example.

File: src/main/java/sg/dex/starfish/impl/remote/RemoteJob.java [hunk 5]
```diff
@@ -37,7 +37,11 @@ public boolean isDone() {
     @Override
 	@SuppressWarnings("unchecked")
 	public synchronized Map<String,Object> pollResult() {
-        if (response != null) return response;
+        if (response != null) {
+            this.status= Status.succeeded;
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/remote/RemoteJob.java]: Is this correct? What if the response indicates a remote failure? Needs good tests.

Also I think "pollResult()" needs to do a remote HTTP call to get the latest status?

File: src/test/java/sg/dex/starfish/impl/operations/TestMemoryOperations.java [hunk 6]
```diff
@@ -79,9 +75,12 @@ public void testPrimeAsync() throws IOException {
 
         Map<String, Object> test = new HashMap<>();
         test.put("input", "10");
+
         Job job = memoryOperation.invokeAsync(test);
+        assertEquals(Job.Status.running,job.getStatus());
```
Comment (mikera) [inline src/test/java/sg/dex/starfish/impl/operations/TestMemoryOperations.java]: Race condition? How are you ensuring that the Job has not already completed?

**Review by mikera — CHANGES_REQUESTED** [review]

File: src/main/java/sg/dex/starfish/Job.java (hunk 1 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/starfish/Job.java]: updated

File: src/main/java/sg/dex/starfish/impl/memory/MemoryJob.java (hunk 3 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/starfish/impl/memory/MemoryJob.java]: updated the code

**Pull request merged**